app/services/experiment_control_service.rb

```diff
@@ -0,0 +1,94 @@
+class InvalidExperiment < StandardError; end
+
+class ExperimentControlService
+  class << self
+    def start_current_patient_experiment(name, days_til_start, days_til_end)
+      raise ArgumentError, "Start date must be before end date" if days_til_end < days_til_start
+
+      existing_experiment = Experimentation::Experiment.where(experiment_type: "current_patient_reminder", state: ["selecting", "live"])
+      raise InvalidExperiment, "A current patient experiment is already in progress" if existing_experiment.any?
+
+      experiment = Experimentation::Experiment.find_by!(name: name, experiment_type: "current_patient_reminder")
+      experiment_start = days_til_start.days.from_now.beginning_of_day
+      experiment_end = days_til_end.days.from_now.end_of_day
+
+      experiment.state_selecting!
+
+      experiment_patients = patient_pool
+        .joins(:appointments)
+        .where("appointments.status = ?", "scheduled")
+        .where("appointments.scheduled_date BETWEEN ? AND ?", experiment_start, experiment_end)
+        .order(Arel.sql("random()"))
+
+      experiment_patients.each do |patient|
+        group = experiment.group_for(patient.id)
+        appointment = patient.appointments
+          .where(status: "scheduled")
+          .where("appointments.scheduled_date BETWEEN ? AND ?", experiment_start, experiment_end).first
+        schedule_reminders(patient, appointment, experiment, group, appointment.scheduled_date)
```

Comment:
We are passing a lot of state around here....could we get this down to perhaps just `patient, appointment, group` ? I think we can derive the rest from that...

Reply:
We can definitely derive experiment, but schedule_date does not come from appointment for stale patients. But that could change depending on another open question. In a comment above, I had said that grabbing the last appointment for a stale patient might not be reliable and that it might be better to create an appointment for this purpose. If we created an appointment for the date the patient's reminders should start, we would not need an additional argument for schedule date.